src/CommunityToolkit.Aspire.Hosting.SqlDatabaseProjects/SqlProjectPublishService.cs

```diff
@@ -0,0 +1,67 @@
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Eventing;
+using Microsoft.Extensions.Logging;
+
+namespace CommunityToolkit.Aspire.Hosting.SqlDatabaseProjects;
+
+internal class SqlProjectPublishService
+{
+    private readonly IDacpacDeployer _deployer;
+    private readonly ResourceLoggerService _resourceLoggerService;
+    private readonly ResourceNotificationService _resourceNotificationService;
+    private readonly IDistributedApplicationEventing _eventing;
+    private readonly IServiceProvider _serviceProvider;
+
+    public SqlProjectPublishService(IDacpacDeployer deployer, ResourceLoggerService resourceLoggerService,
+        ResourceNotificationService resourceNotificationService, IDistributedApplicationEventing eventing, IServiceProvider serviceProvider)
+    {
+        _deployer = deployer ?? throw new ArgumentNullException(nameof(deployer));
+        _resourceLoggerService = resourceLoggerService ?? throw new ArgumentNullException(nameof(resourceLoggerService));
+        _resourceNotificationService = resourceNotificationService ?? throw new ArgumentNullException(nameof(resourceNotificationService));
+        _eventing = eventing ?? throw new ArgumentNullException(nameof(eventing));
+        _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+    }
+
+    public async Task PublishSqlProject(SqlProjectResource sqlProject, SqlServerDatabaseResource target, CancellationToken cancellationToken)
```

Comment:
I think it'd be better to use `KnownResourceStates.FailedToStart` rather than `Failed` when there are errors in here, as then we're more consistent with how Aspire reports statuses.